README.md
# DiscoAuth
automatically follows through with the requests for Discord OAuth

[![Documentation](https://img.shields.io/badge/docs-passing-light_green?style=for-the-badge&logo=readthedocs)](https://discoauth.rtfd.io)
[![Codecov](https://img.shields.io/codecov/c/gh/disoauth/DiscoAuth?style=for-the-badge&logo=codecov)](https://codecov.io/gh/disoauth/DiscoAuth)

***

## Features
- Makes a Authorization Url automatically, using the client ID, redirect_uri, and the scopes you want
- Can get:
  - The access token
  - The user
  - The user's guilds
  - A specific guild, using it's ID

It also makes bot auth url's

***

## What I'm working on for version 3.0.0

*P.S.* The version number *(Major.Minor)* following the functions and models show which versions I intend to add them

- [ ] Add functions and models for User and Guild
  - [ ] User
    - [ ] Functions
      <details>
      <summary>User Functions</summary>
        
      - [x] Get User - v2.1
      - [x] Modify Current User - v2.1
      - [ ] Get Current Guild Member - v2.3
      - [x] Leave Guild - v2.1
      - [x] Create DM - v2.1
      - [x] Create Group DM - v2.1
      - [ ] Get Current User Connections - v2.1
      - [ ] Get Current User Application Role Connection - v2.1
      - [ ] Update Current User Application Role Connection - v2.1
  - [ ] Guild
    - [ ] Functions
      <details>
      <summary>Check The Guild Functions (It's long)</summary>
        
      - [ ] Create Guild - v2.2
      - [ ] Get Guild Preview - v2.2
      - [ ] Modify Guild - v2.2
      - [ ] Delete Guild - v2.2
      - [ ] Get Guild Channels - v2.4
      - [ ] Create Guild Channels - v2.4
      - [ ] Modify Guild Channels Positions - v2.4
      - [ ] List Active Guild Threads - v2.2
      - [ ] Get Guild Member - v2.3
      - [ ] List Guild Members - v2.3
      - [ ] Search Guild Members - v2.3
      - [ ] Add Guild Member - v2.3
      - [ ] Modify Guild Member - v2.3
      - [ ] Modify Current Member - v2.3
      - [ ] Modify Current User Nick - v2.3
      - [ ] Add Guild Member Role - v2.4
      - [ ] Remove Guild Member Role - v2.4
      - [ ] Remove Guild Member - v2.3
      - [ ] Get Guild Bans - v2.2
      - [ ] Get Guild Ban - v2.2
      - [ ] Create Guild Ban - v2.2
      - [ ] Remove Guild Ban - v2.2
      - [ ] Get Guild Roles - v2.5
      - [ ] Create Guild Role - v2.5
      - [ ] Modify Guild Role Positions - v2.5
      - [ ] Modify Guild Role - v2.5
      - [ ] Modify Guild MFA Level - v2.2
      - [ ] Delete Guild Role - v2.5
      - [ ] Get Guild Prune Count - v2.3
      - [ ] Begin Guild Prune - v2.3
      - [ ] Get Guild Voice Regions - v2.6
      - [ ] Get Guild Invites - v2.2
      - [ ] Get Guild Integrations - v2.6
      - [ ] Delete Guild Integration
      - [ ] Get Guild Widget Settings
      - [ ] Modify Guild Widget
      - [ ] Get Guild Widget
      - [ ] Get Guild Vanity URL
      - [ ] Get Guild Widget Image
      - [ ] Get Guild Welcome Screen
      - [ ] Modify Guild Welcome Screen
      - [ ] Get Guild Onboarding
      - [ ] Modify Guild Onboarding
      - [ ] Modify Current User Voice State
      - [ ] Modify User Voice State
      </details>
    - [ ] Models (Like channel, and role models)
      - [ ] Guild Preview Model - v2.2
      - [ ] Guild Member Model - v2.3
      - [ ] Channel Model - v2.4
      - [ ] Role Model - v2.5
      - [ ] Integration Model - v2.7
      - [ ] welcome Screen Model - v2.8
      - [ ] Guild Onboarding Model - v2.8
      - [ ] Guild Widget - v2.7
- [ ] Raise `MissingScopes` when trying to use a functions that need extra scopes

***

### Useful Links:

- [License](https://github.com/disoauth/DiscoAuth/blob/main/LICENSE)
- [Contributing & Guidelines](https://github.com/disoauth/DiscoAuth/blob/main/CONTRIBUTING.md)
- [Roadmap](https://github.com/orgs/disoauth/projects/1)

***

## Coverage

[![Coverage](https://codecov.io/gh/disoauth/DiscoAuth/graphs/sunburst.svg?token=0A6DPREED2)](https://codecov.io/gh/disoauth/DiscoAuth)

***

## Contributors

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tbody>
    <tr>
      <td align="center" valign="top" width="14.28%"><a href="https://github.com/Arcader717"><img src="https://avatars.githubusercontent.com/u/134526462?v=4?s=100" width="100px;" alt="Arcader717"/><br /><sub><b>Arcader717</b></sub></a><br /><a href="#code-Arcader717" title="Code">💻</a> <a href="#doc-Arcader717" title="Documentation">📖</a> <a href="#example-Arcader717" title="Examples">💡</a> <a href="#tutorial-Arcader717" title="Tutorials">✅</a></td>
    </tr>
  </tbody>
</table>

<!-- markdownlint-restore -->
<!-- prettier-ignore-end -->

<!-- ALL-CONTRIBUTORS-LIST:END -->

***

## License

This project is under the [MIT License](https://en.wikipedia.org/wiki/MIT_License)


docs/CHANGELOG.rst
=========
Changelog
=========

All notable changes to this project will be documented in this file.

The format is based on `Keep a Changelog <https://keepachangelog.com/en/1.0.0/>`_,
and this project adheres to `Semantic Versioning <https://semver.org/spec/v2.0.0.html>`_.

[`Unreleased`_]
===============

Added
-----

* Get User Function, using a user id, added to the :meth:`discord.user.fetch` method of the :class:`discord.user` class
* Modify Current User
* Get Current Guild Member
* Leave Guild
* Create DM
* Create Guild DM

Fixed
-----

* The documentation in the :doc:`permissions` page

[`2.0.0`_] - November 13, 2023
==============================

`Diff`_ for 2.0.0 and 1.4.0 

Not much difference between 2.0.0 and 2.0.0-rc.1

Changed
-------

* The name of the project to **`discoauth`**

[`2.0.0-rc.1`_] - November 13, 2023
===================================

Used to release the final version to Async-DisOAuth

Added
-----

- [`#70 <https://github.com/disoauth/DiscoAuth/pull/70>`_]A :code:`deprecated` folder to hold the deprecated functions 
- New subclass of :class:`discord`, :class:`discord.guild`, to help get specific guilds instead of user guilds

Deprecated
----------

- :class:`AuthUrl`, replaced with :class:`auth`

  - :meth:`AuthUrl.makeUrl`, replaced with :meth:`auth.url`

- :class:`discordApi`, replaced with :class:`discord`

  - :meth:`discordApi.accessToken`, replaced with :meth:`discord.token`
  - :class:`discordApi.User`, replaced with :class:`discord.user` 

    - :meth:`discordApi.User.get_current_user`, replaced with :meth:`discord.user.fetch`
    - :meth:`discordApi.User.get_user_guilds`, replaced with :meth:`discord.user.guilds`
    - :meth:`discordApi.User.get_guild`, replaced with :meth:`discord.guild.fetch` (replace the :class:`discord.user` class with :class:`discord.guild`, to get this to work)


[`1.4.0`_] - November 13, 2023
==============================

Added
-----

- [`#53 <https://github.com/disoauth/DiscoAuth/pull/53>`_] CLI Update (The big point of the next version)
- [`#59 <https://github.com/disoauth/DiscoAuth/pull/59>`_] Tests (Kinda realized that they are really helpful)

Changed
-------

* Moved the docs from ``../docs/source`` to ``../docs``

Fixed
-----

* [`#55 <https://github.com/disoauth/DiscoAuth/pull/55>`_] The scope of the auth url

Removed
-------

* The markdown documentation, for MKDocs, which I am no longer using

[`1.3.0`_] - November 8, 2023
=============================

Added
-----

* [`#33 <https://github.com/disoauth/DiscoAuth/pull/33>`_] :class:`bot` Class, for bot auth

 * :class:`permissions` class, to help with the bot auth

* Documentation:

 * New Contributing Page
 * New deprecate page, for the upcoming renaming of ``DisOAuth`` to ``discoauth``

[`1.2.0`_] - November 6, 2023
=============================

Added
-----

* The all-contributors bot
* A CONTRIBUTING.md
* Guild Object
* A new method in the :class:`User` class, that returns a guild object
* A :code:`with_count` parameter to the methods that return a guild object
* documentation for the guild object

Changed
-------

* The repo name from ``Arcader717/DiscordOAuth2`` to ``disoauth/DiscoAuth``

[`1.1.5`_] - October 31, 2023
=============================

Fixed
-----

* the scope of the auth url

[`1.1.5b1`_] - October 30, 2023
===============================

Fixed
-----

* The scope of the auth url

[`1.1.4`_] - October 30, 2023 
=============================

Fixed
-----

* The scope of the auth url

[`1.1.3`_] - October 27, 2023
=============================

Changed
-------

* [`#21 <https://github.com/disoauth/DiscoAuth/pull/21>`_] Moved the user model to :code:`models.py`

[`1.1.2`_] - October 25, 2023
=============================

Added
-------

* Added the :class:`UserObj` to the documentation


[`1.1.1`_] - October 25, 2023
=============================

Changed
-------

* The goals in the README from the v1.0.0 goals to v2.0.0 goals

[`1.1.0`_] - October 25, 2023
=============================

Added
-----

* Documentation, using `Read The Docs <https://readthedocs.io>`_ and `Sphinx <https://www.sphinx-doc.org/en/master/usage/index.html>`_
* [`#18 <https://github.com/disoauth/DiscoAuth/pull/18>`_] A model for user data

[`1.0.1`_] - September 22, 2023
===============================

Nothing important happened for this update

[`1.0.0`_] - September 22, 2023
===============================

Added
-----

* [`#7 <https://github.com/disoauth/DiscoAuth/pull/7>`_] The subclass, :class:`User` to :class:`discordApi`
* A method in the new :class:`User` class to get user data
* A method to make the auth url in the AuthUrl class

Fixed
-----

* The state generation

[`0.2.1`_] - September 18, 2023
===============================

Changed
-------

* Renamed Package from :code:`DisOAuth` to :code:`Async-DisOAuth` to accomodate for the Sync version of :code:`DisOAuth`

  * (Later, the Sync version would be discontinued)

[`0.2.0`_] - September 15, 2023
===================================================================================

Added
-----

* The :class:`discordApi` class
* The :meth:`discordApi.accessToken` method, to get the access token of a user
* A method of :class:`discordApi` to check the app's auth info
* MIT License

[`0.1.4`_] - September 13, 2023
===================================================================================

Changed
-------

* Package renamed from :code:`DiscordOAuth2` to :code:`DisOAuth`

[`0.1.3`_] - September 13, 2023
===================================================================================

Changed
-------

* Package renamed from :code:`DiscordOAuth-Helper` to :code:`DiscordOAuth2` 

[`0.1.2`_] - September 13, 2023
===================================================================================

Added
-----

* The authors in pyproject.toml [`Commit <https://github.com/disoauth/DiscoAuth/commit/dd673466ba882fa6dca4bd1dbfa793158878b2d3>`_]

[`0.1.1`_] - September 13, 2023

src/DisOAuth/common.py
import json
from random import SystemRandom
from typing import List

import requests

_UNICODE_ASCII_CHARACTER_SET = ('a',
                                'b',
                                'c',
                                'd',
                                'e',
                                'f',
                                'g',
                                'h',
                                'i',
                                'j',
                                'k',
                                'l',
                                'm',
                                'n',
                                'o',
                                'p',
                                'q',
                                'r',
                                's',
                                't',
                                'u',
                                'v',
                                'w',
                                'x',
                                'y',
                                'z',
                                'A',
                                'B',
                                'C',
                                'D',
                                'E',
                                'F',
                                'G',
                                'H',
                                'I',
                                'J',
                                'K',
                                'L',
                                'M',
                                'N',
                                'O',
                                'P',
                                'Q',
                                'R',
                                'S',
                                'T',
                                'U',
                                'V',
                                'W',
                                'X',
                                'Y',
                                'Z',
                                '0',
                                '1',
                                '2',
                                '3',
                                '4',
                                '5',
                                '6',
                                '7',
                                '8',
                                '9')


async def generate_token(length=30, chars=_UNICODE_ASCII_CHARACTER_SET):
    rand = SystemRandom()
    return ''.join(rand.choice(chars) for x in range(length))


async def getToken(code: str,
             scope: list[str],
             redirect_uri: str,
             client_id: str,
             client_secret: str) -> dict[str, str]:
    url = "https://discord.com/api/oauth2/token"
    scopeStr = " ".join(scope)
    rDict: dict[str, str] = {"error": "nothing loaded"}
    data = {
        'grant_type': 'authorization_code',
        'code': code,
        'client_id': client_id,
        'client_secret': client_secret,
        'redirect_uri': redirect_uri
    }
    headers = {'Content-Type': 'application/x-www-form-urlencoded'}
    r = requests.post(url=url,
                      data=data,
                      headers=headers)
    return r.json()  # returns the json file

async def joinUrl(clientID: str, scopes: str, redirectUri: str, state: str) -> str:
    baseUrl = "https://discord.com/oauth2/authorize?"
    urlList = [baseUrl,
               "response_type=code",
               "&",
               "client_id=",
               clientID,
               "&",
               "scope=",
               scopes,
               "&",
               "state=",
               state,
               "&",
               "redirect_uri=",
               redirectUri]
    url = "".join(urlList)
    return url

bot_perms = {
  'CREATE_INSTANT_INVITE': 0x1,
  'KICK_MEMBERS': 0x2,
  'BAN_MEMBERS': 0x4,
  'ADMINISTRATOR': 0x8,
  'MANAGE_CHANNELS': 0x10,
  'MANAGE_GUILD': 0x20,
  'ADD_REACTIONS': 0x40,
  'VIEW_AUDIT_LOG': 0x80,
  'PRIORITY_SPEAKER': 0x100,
  'STREAM': 0x200,
  'VIEW_CHANNEL': 0x400,
  'SEND_MESSAGES': 0x800,
  'SEND_TTS_MESSAGES': 0x1000,
  'MANAGE_MESSAGES': 0x2000,
  'EMBED_LINKS': 0x4000,
  'ATTACH_FILES': 0x8000,
  'READ_MESSAGE_HISTORY': 0x10000,
  'MENTION_EVERYONE': 0x20000,
  'USE_EXTERNAL_EMOJIS': 0x40000,
  'VIEW_GUILD_INSIGHTS': 0x80000,
  'CONNECT': 0x100000,
  'SPEAK': 0x200000,
  'MUTE_MEMBERS': 0x400000,
  'DEAFEN_MEMBERS': 0x800000,
  'MOVE_MEMBERS': 0x1000000,
  'USE_VAD': 0x2000000,
  'CHANGE_NICKNAME': 0x4000000,
  'MANAGE_NICKNAMES': 0x8000000,
  'MANAGE_ROLES': 0x10000000,
  'MANAGE_WEBHOOKS': 0x20000000,
  'MANAGE_GUILD_EXPRESSIONS': 0x40000000,
  'USE_APPLICATION_COMMANDS': 0x80000000,
  'REQUEST_TO_SPEAK': 0x100000000,
  'MANAGE_EVENTS': 0x200000000,
  'MANAGE_THREADS': 0x400000000,
  'CREATE_PUBLIC_THREADS': 0x800000000,
  'CREATE_PRIVATE_THREADS': 0x1000000000,
  'USE_EXTERNAL_STICKER': 0x2000000000,
  'SEND_MESSAGES_IN_THREADS': 0x4000000000,
  'USE_EMBEDDED_ACTIVITIES': 0x8000000000,
  'MODERATE_MEMBERS': 0x10000000000,
  'VIEW_CREATOR_MONETIZATION_ANALYTICS': 0x20000000000,
  'USE_SOUNDBOARD': 0x40000000000,
  'USE_EXTERNAL_SOUNDS': 0x200000000000,
  'SEND_VOICE_MESSAGES': 0x400000000000
}

bot_perms_key = {
      0:'CREATE_INSTANT_INVITE',
      1:'KICK_MEMBERS',
      2:'BAN_MEMBERS',
      3:'ADMINISTRATOR',
      4:'MANAGE_CHANNELS',
      5:'MANAGE_GUILD',
      6:'ADD_REACTIONS',
      7:'VIEW_AUDIT_LOG',
      8:'PRIORITY_SPEAKER',
      9:'STREAM',
      10:'VIEW_CHANNEL',
      11:'SEND_MESSAGES',
      12:'SEND_TTS_MESSAGES',
      13:'MANAGE_MESSAGES',
      14:'EMBED_LINKS',
      15:'ATTACH_FILES',
      16:'READ_MESSAGE_HISTORY',
      17:'MENTION_EVERYONE',
      18:'USE_EXTERNAL_EMOJIS',
      19:'VIEW_GUILD_INSIGHTS',
      20:'CONNECT',
      21:'SPEAK',
      22:'MUTE_MEMBERS',
      23:'DEAFEN_MEMBERS',
      24:'MOVE_MEMBERS',
      25:'USE_VAD',
      26:'CHANGE_NICKNAME',
      27:'MANAGE_NICKNAMES',
      28:'MANAGE_ROLES',
      29:'MANAGE_WEBHOOKS',
      30:'MANAGE_GUILD_EXPRESSIONS',
      31:'USE_APPLICATION_COMMANDS',
      32:'REQUEST_TO_SPEAK',
      33:'MANAGE_EVENTS',
      34:'MANAGE_THREADS',
      35:'CREATE_PUBLIC_THREADS',
      36:'CREATE_PRIVATE_THREADS',
      37:'USE_EXTERNAL_STICKER',
      38:'SEND_MESSAGES_IN_THREADS',
      39:'USE_EMBEDDED_ACTIVITIES',
      40:'MODERATE_MEMBERS',
      41:'VIEW_CREATOR_MONETIZATION_ANALYTICS',
      42:'USE_SOUNDBOARD',
      45:'USE_EXTERNAL_SOUNDS',
      46:'SEND_VOICE_MESSAGES'
}

async def permsByList(perm_list: List[int]) -> int:
    perms = 0x0
    for perm in perm_list:
        perms |= bot_perms[bot_perms_key[perm]]
    return perms

async def htmlEncode(str):
    str1 = str.replace(" ", "%20")
    str2 = str1.replace(":", "%3A")
    str3 = str2.replace("/", "%2F")
    return str3
